pkg/porter/build.go

```diff
@@ -20,24 +21,61 @@ type BuildProvider interface {
 
 type BuildOptions struct {
 	contextOptions
+	metadataOpts
 	NoLint bool
 }
 
+func (o BuildOptions) Validate() error {
+	if o.Name != "" {
+		// TODO: What sort of validation, if any, do we want on the bundle name?
+		// Originally I had used the *installation* validation from cnab-go,
+		// but that is most likely more restrictive than what we want (no spaces allowed, etc.)
+	}
+
+	if o.Version != "" {
+		if _, err := semver.NewVersion(o.Version); err != nil {
+			return errors.Wrapf(err, "invalid bundle version %q.  Cannot be parsed as semver", o.Version)
+		}
+	}
+
+	return nil
+}
+
 func (p *Porter) Build(opts BuildOptions) error {
 	opts.Apply(p.Context)
 
-	err := p.LoadManifest()
-	if err != nil {
+	if err := opts.Validate(); err != nil {
 		return err
 	}
 
+	if err := p.generateInternalManifest(opts.metadataOpts); err != nil {
+		return errors.Wrap(err, "unable to generate manifest")
+	}
+
+	// Publish may invoke this method and the manifest will already be
+	// populated.  Only load if still empty.
+	if p.Manifest == nil {
```

Comment:
I am concerned that this if check here could cause line 51 to generate a different manifest than what is already loaded. It may be safer long term to unconditionally load the manifest that we just generated.

Especially since we came into this method from Publish deciding the manifest it has is out of date?

Reply:
Yes, this is a tricky area.  This check was added in response to the following flow in particular:

1. `porter publish --tag diffreg/diffbundle:v0.1.0` -> Currently, we update the bundle reference and invocation image on the manfiest from this supplied tag, which triggers a rebuild owing to the updated manifest fields.
2. `porter build` is therefore invoked, however, currently, we don't pass along this tag value.  The updated manifest fields in memory are then lost when we call `p.LoadManifestFrom(build.LOCAL_MANIFEST)` here.
3. `porter publish --tag diffreg/diffbundle:v0.1.0` continues, albeit with the reloaded manifest and values for image and reference from `build.LOCAL_MANIFEST` and not the correct values derived from the supplied `tag`...

A few options we might entertain (also eager to consider any others!):

- Start passing the `tag` value along whenever publish triggers a build, so that build can use this value to make sure the name, version and 
- Update the latter (#3) logic in publish to use bundle ref and image values parsed from the supplied `tag` value when publishing the bundle and image to the registry, as opposed to using what is on the manifest itself.

Reply:
Let me restate to make sure I understand. Is this what your change does? How close is your PR to this?

1. The bundle is originally built with `name: mybun`, `version: 1.0.0`, `registry: someplace/`. The invocation image name stored in the bundle.json is `someplace/mybun:v1.0.0`.
2. The bundle is published to an alternate location that doesn't match the existing name + version + registry. For example, `someplace/mybun:latest` or `anotherplace/mybun:v1.0.0`.
3. On publish we are parsing that tag and *updating the in-memory manifest* to match the values from tag (name, version and registry). That normally triggers a rebuild. We shouldn't rewrite the rendered manifest to disk.

I'm going to stop at this point. Does this match what you are thinking? We don't want to rebuild in this scenario.

Instead we want to push a bundle to a different location without modifying the invocation image through the use of the relocation mapping file. We push the original invocation image to the new location, map the original invocation image to the new one, and give the mapping file to cnab-to-oci.

Essentially specifying `porter publish --reference/tag` should not cause the metadata of the bundle to change. It only says "push it to someplace other than the default generated from the metadata".

\* I am ignoring https://github.com/getporter/porter/discussions/1360 for now.

Reply:
Yup, you've got it.  As mentioned, currently, 3. _does_ trigger a rebuild, but we'd like to soon move to utilizing the relocation mapping file such that a rebuild is not needed.  Ideally we can prioritize this as a follow-up.

In this PR, I'm attempting to retain the status quo where publishing w/ a different tag/reference correctly triggeres a rebuild... (otherwise, if we don't rebuild -- assuming we don't yet use the relo map -- publish will fail as it can't find the invocation image locally).  Perhaps I can revisit to see if there's a better way to trigger this without sparking a non-trivial refactor...

Reply:
No that's fine! You cleared it up for me.